src/Service/MailTemplateHelper.php

```diff
@@ -0,0 +1,283 @@
+                $placeHolders,
+                array_map(static function ($name) {
+                    $marker = new TextRun();
+                    $marker->addText('${'.$name.'}', ['bgColor' => 'FFFF99']);
+
+                    return $marker;
+                }, $placeHolders)
+            );
+        } else {
+            // Convert entity to array.
+            $data = json_decode($this->serializer->serialize($entity, 'json', ['groups' => ['mail_template']]), true);
+            $values += $this->flatten($data);
+        }
+
+        // Add user info
+        $user = $this->tokenStorage->getToken()?->getUser();
+        if (null !== $user && $user instanceof User) {
+            $values += $this->flatten([
+                'user' => [
+                    'email' => $user->getEmail(),
+                    'name' => $user->getName(),
+                ],
+            ]);
+        }
+
+        // Add some default values.
+        $values['now'] = (new \DateTimeImmutable('now'))->format(\DateTimeImmutable::ATOM);
+        $values['today'] = (new \DateTimeImmutable('today'))->format(\DateTimeImmutable::ATOM);
+
+        foreach ($placeHolders as $placeHolder) {
+            // Handle placeholders on the form «key»:«format», e.g.
+            //   case.inspection_date:dd/mm/yyyy
+            if (preg_match('/^(?P<key>[^:]+):(?P<format>.+)$/', $placeHolder, $matches)) {
+                [, $key, $format] = $matches;
+                if (isset($values[$key])) {
+                    $values[$placeHolder] = $this->formatValue($values[$key], $format);
+                }
+            }
+        }
+
+        // Set empty string values for all template variables without a value.
+        $values += array_map(static function ($count) { return ''; }, $templateProcessor->getVariableCount());
+
+        return $values;
+    }
+
+    /**
+     * Format a value.
+     */
+    private function formatValue(string $value, string $format): string
+    {
+        // Try to format the value as a date(time).
+        try {
+            $date = new \DateTimeImmutable($value);
+            $formatter = new \IntlDateFormatter(\Locale::getDefault(), \IntlDateFormatter::NONE, \IntlDateFormatter::NONE);
+            // https://unicode-org.github.io/icu/userguide/format_parse/datetime/#datetime-format-syntax
+            $formatter->setPattern($format);
+
+            return $formatter->format($date);
+        } catch (\Exception $exception) {
+        }
+
+        return $value;
+    }
+
+    private function setTemplateValue(string $name, $value, TemplateProcessor $templateProcessor)
+    {
+        if ($value instanceof AbstractElement) {
+            $count = 0;
+            // TemplateProcessor::setComplexValue() only replaces one (the first) occurrence.
+            while (isset($templateProcessor->getVariableCount()[$name]) && $count < 10) {
+                $templateProcessor->setComplexValue($name, $value);
+                ++$count;
+            }
```

Comment:
It is not very clear to me exactly what is happening here, a helping comment would be nice.